File: backend/pkg/handlers/web/clickRage.go
package web

import (
	"encoding/json"
	"log"

	. "openreplay/backend/pkg/messages"
)

const MaxTimeDiff = 300
const MinClicksInARow = 3

type ClickRageDetector struct {
	lastTimestamp        uint64
	lastLabel            string
	lastSelector         string
	firstInARawTimestamp uint64
	firstInARawMessageId uint64
	countsInARow         int
	url                  string
}

func (crd *ClickRageDetector) reset() {
	crd.lastTimestamp = 0
	crd.lastLabel = ""
	crd.lastSelector = ""
	crd.firstInARawTimestamp = 0
	crd.firstInARawMessageId = 0
	crd.countsInARow = 0
	crd.url = ""
}

func (crd *ClickRageDetector) createPayload() string {
	p, err := json.Marshal(struct{ Count int }{crd.countsInARow})
	if err != nil {
		log.Printf("can't marshal ClickRage payload to json: %s", err)
		return ""
	}
	return string(p)
}

func (crd *ClickRageDetector) Build() Message {
	defer crd.reset()
	if crd.countsInARow < MinClicksInARow {
		return nil
	}
	return &IssueEvent{
		Type:          "click_rage",
		ContextString: crd.lastLabel,
		Payload:       crd.createPayload(),
		Timestamp:     crd.firstInARawTimestamp,
		MessageID:     crd.firstInARawMessageId,
		URL:           crd.url,
		Context:       crd.lastSelector, // hack to pass selector to db (tags filter)
	}
}

func (crd *ClickRageDetector) Handle(message Message, timestamp uint64) Message {
	switch msg := message.(type) {
	case *MouseClick:
		// Set click url
		if crd.url == "" && msg.Url != "" {
			crd.url = msg.Url
		}
		// Click on different object -> build if we can and reset the builder
		if msg.Label == "" {
			return crd.Build()
		}
		// Update builder with last information
		if crd.lastLabel == msg.Label && timestamp-crd.lastTimestamp < MaxTimeDiff {
			crd.lastTimestamp = timestamp
			crd.countsInARow += 1
			return nil
		}
		// Try to build event
		event := crd.Build()
		// Use current message as init values for new event
		crd.lastTimestamp = timestamp
		crd.lastLabel = msg.Label
		crd.lastSelector = msg.Selector
		crd.firstInARawTimestamp = timestamp
		crd.firstInARawMessageId = message.MsgID()
		crd.countsInARow = 1
		if crd.url == "" && msg.Url != "" {
			crd.url = msg.Url
		}
		return event
	}
	return nil
}

File: tracker/tracker/src/webworker/detectors/clickRage.ts
import Message, {
  Type,
  MouseClick,
} from '../../common/messages.gen.js'
import type { Detector, ReportIssue } from './types.js'

/**
 * Port of backend/pkg/handlers/web/clickRage.go.
 *
 * Fires when the same element is clicked >= MIN_CLICKS_IN_A_ROW times, each
 * within MAX_TIME_DIFF ms of the previous one.
 *
 * URL is stamped centrally by the Detectors layer (latest SetPageLocation), so
 * this detector doesn't track it.
 */

const MAX_TIME_DIFF = 300
const MIN_CLICKS_IN_A_ROW = 3

export default class ClickRageDetector implements Detector {
  private lastTimestamp = 0
  private lastLabel = ''
  private lastSelector = ''
  private firstInARowTimestamp = 0
  private firstInARowMessageId = 0
  private countsInARow = 0

  constructor(private readonly report: ReportIssue) {}

  private reset(): void {
    this.lastTimestamp = 0
    this.lastLabel = ''
    this.lastSelector = ''
    this.firstInARowTimestamp = 0
    this.firstInARowMessageId = 0
    this.countsInARow = 0
  }

  private build(): void {
    const count = this.countsInARow
    const label = this.lastLabel
    const selector = this.lastSelector
    const firstTs = this.firstInARowTimestamp
    const messageId = this.firstInARowMessageId
    // matches Go `defer crd.reset()`
    this.reset()
    if (count < MIN_CLICKS_IN_A_ROW) {
      return
    }
    this.report({
      type: 'click_rage',
      contextString: label,
      context: selector, // used by the tags filter
      payload: JSON.stringify({ Count: count }),
      timestamp: firstTs,
      messageId,
    })
  }

  handle(message: Message, index: number, timestamp: number): void {
    const type = message[0]
    if (type !== Type.MouseClick) {
      return
    }
    const msg = message as MouseClick
    const label = msg[3]
    const selector = msg[4]
    if (label === '') {
      this.build()
      return
    }
    if (this.lastLabel === label && timestamp - this.lastTimestamp < MAX_TIME_DIFF) {
      this.lastTimestamp = timestamp
      this.countsInARow += 1
      return
    }
    this.build()
    this.lastTimestamp = timestamp
    this.lastLabel = label
    this.lastSelector = selector
    this.firstInARowTimestamp = timestamp
    this.firstInARowMessageId = index
    this.countsInARow = 1
  }
}
